pex/pex.py

```diff
@@ -617,10 +617,47 @@ def execute_script(self, script_name):
         dist_script = get_script_from_distributions(script_name, dists)
         if not dist_script:
             return "Could not find script {!r} in pex!".format(script_name)
-        TRACER.log("Found script {!r} in {!r}.".format(script_name, dist))
-        return self.execute_content(
-            dist_script.path, dist_script.read_contents(), argv0=script_name
-        )
+        TRACER.log("Found script {!r} in {!r}.".format(script_name, dist_script.dist))
+
+        contents = dist_script.read_contents()
+        shebang = contents.split("\n", 1)[0]
+        try:
+            interpreter = PythonInterpreter.from_shebang(shebang)
+            if interpreter and interpreter.version_string != self.interpreter.version_string:
+                TRACER.log(
+                    "Python script shebang resolved to {shebang_python}, but we will execute it with {pex_python}".format(
+                        shebang_python=interpreter.version_string,
+                        pex_python=self.interpreter.version_string,
+                    ),
+                    V=2,
+                )
+        except PythonInterpreter.InterpreterNotFound as e:
+            TRACER.log(
+                "Python script shebang indicated unknown interpreter: {shebang!r}: {e}, will execute it with {version}.".format(
+                    shebang=shebang, e=e, version=self.interpreter.version_string
+                ),
+                V=2,
+            )
+            interpreter = self.interpreter
+
+        if interpreter:
+            TRACER.log(
+                "Executing python script with {}".format(self.interpreter.version_string), V=2
+            )
+            return self.execute_content(dist_script.path, contents, argv0=script_name)
+        else:
+            TRACER.log("Executing non-python script using shebang: {}".format(shebang), V=2)
+            return self.execute_external(dist_script.path)
+
+    def execute_external(self, cmd):
+        # type: (str) -> Any
+        try:
+            # execute as a shell command, inheriting stdout and stderr
```

Comment:
This comment just restates the code. The only interesting thing to comment on is left out. Why the [risky](https://docs.python.org/3/library/subprocess.html#security-considerations) `shell=True`? Do you know of cases where the non-python script is not extracted / installed with the executable bit set?

Reply:
No, I do not. So, more likely my understanding of how popen and the shell parameter works that is a bit fuzzy.
I read the security implications you linked to a few times over.. a few assumptions etc led me down the wrong path, then ;)

Reply:
Been thinking a bit more about this, however, and wonder what the risk is, in this particular case.
We see it as risky, to execute the first line of a script using the shell, but what about the rest of the file, then?

See if I can illustrate my point:
```bash
#!/usr/bin/env bash

sudo rm -rf /
```

If this was the console_script we're about to execute, regardless of what ill-mannered code we might put at the shebang line, does it really matter? Given what we can do once the script executes?
Feels like I'm missing something here... a chain is no stronger than its weakest link, and all that..

Not saying that we should keep `shell=True`, just wondering if the security risk involved really is applicable in this case.

Reply:
The risk is not really the point here - the point is a standard risky bit of the Popen API is being used without comment _and_ its un-needed to boot (or it should be un-needed). If it is needed, it would be great to find out why and if that's not correctable, add a comment saying why this standard risky bit is being set (ie: the execute bit is not set on all or any scripts so we can't just execute them but need to use the standard system shell as an intermediary). This seems to me similar to using `unsafe` in production rust code - you always want a comment explaining why the `unsafe` is actually safe.

Reply:
Thanks. That makes sense. :)